Use bash -o pipefail for all shell commands in Buildbot

gsnedders opened the pull request:
#### afddd6334cd18be62056c06f07a70c8734d3504b
<pre>
Use bash -o pipefail for all shell commands in Buildbot
<a href="https://bugs.webkit.org/show_bug.cgi?id=276081">https://bugs.webkit.org/show_bug.cgi?id=276081</a>
<a href="https://rdar.apple.com/problem/130908663">rdar://problem/130908663</a>

Reviewed by Elliott Williams, Brianna Fan, and Carlos Alberto Lopez Perez.

We&apos;ve had multiple bugs caused by pipelines exiting cleanly despite
commands within them failing, for example the above linked bug 276081
(where run-webkit-tests exited uncleanly but filter-test-logs exited
cleanly, thus we were able to totally break run-webkit-tests without
noticing), and also bug 271808 (where build-webkit exited uncleanly
but filter-build-webkit exited cleanly, thus we didn&apos;t notice the bot
being in a bogus configuration and no build occurring).

To avoid this, this makes us always pass -o pipefail to the shell we
are creating. This is new in POSIX 2024. On systems using `dash` to
provide `sh` (e.g., Debian and Ubuntu) this is unsupported, as it is
currently only supported in pre-release versions of `dash`. For now,
we use `bash` in its POSIX mode (which is also its default when it is
invoked as `sh`) to try and reduce the risk of bashisms slipping in.

In doing this, we unify ShellMixin to a single implementation in
Shared/steps.py, and import that from both build-webkit-org and
ews-build.

* Tools/CISupport/Shared/steps.py:
(ShellMixin): Update WINDOWS_SHELL_PLATFORMS to match ews-build and build-webkit-org.
(ShellMixin.shell_command): Optionally use pipefail, and when we do, use bash.
* Tools/CISupport/build-webkit-org/steps.py:
(CompileWebKit.start): Use ShellMixin.shell_command, rather than hardcoding /bin/sh.
(RunJavaScriptCoreTests.run): Use ShellMixin.shell_command, rather than hardcoding /bin/sh.
(RunWebDriverTests.run): Use ShellMixin.shell_command, rather than hardcoding /bin/sh.
(PrintConfiguration): Instead of invoking a shell, just run the command directly as buildbot logs the command run.
(ShellMixin): Moved to Shared.
* Tools/CISupport/ews-build/steps.py:
(PrintConfiguration.run): Instead of invoking a shell, just run the command directly as buildbot logs the command run.
(DetermineLandedIdentifier): Run the command with /bin/bash; this is unfortunately not using ShellMixin.
(ShellMixin): Moved to Shared.

Canonical link: <a href="https://commits.webkit.org/294384@main">https://commits.webkit.org/294384@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dc7d402c34f1a9fb14d775827b403f9ba24340bd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100674 "10 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20326 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10625 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105811 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51262 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102715 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20634 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28800 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76662 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33700 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103681 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15837 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90933 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57016 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15646 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8940 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50638 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85556 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9015 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108166 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27792 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20407 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85621 "Passed tests") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/100232 "Passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28155 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87134 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85161 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29860 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7586 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21780 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16539 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27727 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27538 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30856 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29096 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

briannafan reviewed (commented):
Thanks for fixing this!

Just one comment: There's also a `ShellMixin` in `Shared/steps.py`, could you add your changes there? This could also be a good opportunity to consolidate all three and just have one version in `Shared/steps.py`.

emw-apple commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -2068,6 +2082,7 @@ def run(self):
             platform = platform.split('-')[0]
         if platform in ('mac', 'ios', 'visionos', 'tvos', 'watchos', '*'):
             command_list.extend(self.command_list_apple)
+            command_list[5:5] = [self.shell_command('echo TimezoneVers: $(cat /usr/share/zoneinfo/+VERSION)')]
```
nit: I wonder if `echo TimezoneVers:` is really necessary. I don't see anywhere in the codebase parsing for it, and stdout logs show the command as it's being run, so I don't think printing "TimezoneVers:" adds meaningful clarity. 

cc @robert-jenner @JonWBedard as authors of https://commits.webkit.org/259847@main. Perhaps we can remove the `echo` and then treat this like a normal part of the command list?

gsnedders replied:
Given we do indeed seem not to use this anywhere, I've just gone ahead and removed the `echo`.

emw-apple commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -5968,6 +5982,7 @@ def run(self):
             platform = platform.split('-')[0]
         if platform in ('mac', 'ios', 'visionos', 'tvos', 'watchos', '*'):
             command_list.extend(self.command_list_apple)
+            command_list[5:5] = [self.shell_command('echo TimezoneVers: $(cat /usr/share/zoneinfo/+VERSION)')]
```
ditto

clopez reviewed (approved):
Looks good to me.
This should work on Linux (GTK/WPE/JSCOnly) bots.

gsnedders committed:
Use bash -o pipefail for all shell commands in Buildbot
https://bugs.webkit.org/show_bug.cgi?id=276081
rdar://problem/130908663

Reviewed by Elliott Williams, Brianna Fan, and Carlos Alberto Lopez Perez.

We've had multiple bugs caused by pipelines exiting cleanly despite
commands within them failing, for example the above linked bug 276081
(where run-webkit-tests exited uncleanly but filter-test-logs exited
cleanly, thus we were able to totally break run-webkit-tests without
noticing), and also bug 271808 (where build-webkit exited uncleanly
but filter-build-webkit exited cleanly, thus we didn't notice the bot
being in a bogus configuration and no build occurring).

To avoid this, this makes us always pass -o pipefail to the shell we
are creating. This is new in POSIX 2024. On systems using `dash` to
provide `sh` (e.g., Debian and Ubuntu) this is unsupported, as it is
currently only supported in pre-release versions of `dash`. For now,
we use `bash` in its POSIX mode (which is also its default when it is
invoked as `sh`) to try and reduce the risk of bashisms slipping in.

In doing this, we unify ShellMixin to a single implementation in
Shared/steps.py, and import that from both build-webkit-org and
ews-build.

* Tools/CISupport/Shared/steps.py:
(ShellMixin): Update WINDOWS_SHELL_PLATFORMS to match ews-build and build-webkit-org.
(ShellMixin.shell_command): Optionally use pipefail, and when we do, use bash.
* Tools/CISupport/build-webkit-org/steps.py:
(CompileWebKit.start): Use ShellMixin.shell_command, rather than hardcoding /bin/sh.
(RunJavaScriptCoreTests.run): Use ShellMixin.shell_command, rather than hardcoding /bin/sh.
(RunWebDriverTests.run): Use ShellMixin.shell_command, rather than hardcoding /bin/sh.
(PrintConfiguration): Instead of invoking a shell, just run the command directly as buildbot logs the command run.
(ShellMixin): Moved to Shared.
* Tools/CISupport/ews-build/steps.py:
(PrintConfiguration.run): Instead of invoking a shell, just run the command directly as buildbot logs the command run.
(DetermineLandedIdentifier): Run the command with /bin/bash; this is unfortunately not using ShellMixin.
(ShellMixin): Moved to Shared.

Canonical link: https://commits.webkit.org/294384@main
